This is a proposal for updates to the "webhooks" interface

KevinOConnor opened the pull request:
@Arksine - as discussed, I'm putting together a proposal for some changes to the "webhooks" interface.

The primary change is to convert the interface to a format that is more similar to json-rpc (though it does not stringently follow json-rpc).  So, requests now look like: `{"id":14, "method": "api_server/info", "params": {}}`.  The `id` and `params` fields are optional.  Responses now look like `{"id": 14, "result": {}}`.

This branch also adds a new `./scripts/whconsole.py` script that one can use to send in queries to the klippy server.  So, one can send in the above command and see the response.

Another notable change is that the server never sends unsolicited messages to the client.  The client must subscribe to get_status() updates and/or gcode output messages.  Such a subscription might look like: `{"id":101, "method": "subscribe_gcode_output", "params": {"response_template": {"method": "process_gcode_response"}}}`.  On each subscription, one can specify a "response_template" - that template is used to format the message ultimately sent to the client.  For example, after the above, the client might receive: `{"method": "process_gcode_response", "params": {"response": "// ECHO hello=world"}}`

Along with this change, the register_static_path() method was removed - instead, most of that info is now available via the "api_server/info" endpoint.

Also, I propose that we rename "webhooks.py" to "api_server.py" and prefix all endpoints defined in api_server.py with an "api_server/" prefix.

As above, this is just a proposal.  Let me know what you think.

Separately, I think it would be useful to update the get_status() subscription mechanism to support per-client subscriptions.  I also think it would be nice if only deltas were sent after the first update.  That is, if a field like gcode.gcode_position hasn't changed, then no need to send it every update.  If there's interest in that, I can put together a prototype.

-Kevin

Arksine commented:
The proposal looks good to me.  I do have a few comments.

>Also, I propose that we rename "webhooks.py" to "api_server.py" and prefix all endpoints defined in api_server.py with an "api_server/" prefix.

I think this makes sense from Klipper's perspective and we should do this.  When procedurally generating APIs in Moonraker I'll probably need to remove the `api_server` portion, as it makes for Web APIs that are a bit confusing.    For example,  `api_server/objects_subscription` would resolve to a `post_printer_api_server_objects_subscription` websocket API call.

Also, do you think it would be a good idea to have a specific format for rpc handlers similar to gcode handlers?  For example, the info handler could be:
```python
def _rpc_query_INFO(self, web_request):
```
The handler for adding a new subscription could be:
```python
def _rpc_OBJECTS_SUBSCRIPTION(self, web_request):
```

All "queries" would be prefixed with `_rpc_query`, whereas anything that performs an action/mutation is just prefixed with `_rpc`.  We could send the function name along with the registered "method" name in `list_endpoints`, then moonraker could use that info to reject invalid requests.  Pehaps we could shorten `_rpc_query` to `_rpcq` as well.

I think we can remove `api_server/objects_list_subscription` and instead just return the current subscription data when the user adds a subscription.   We can also remove any endpoint that just does nothing but call a gcode as was your original instinct.  

> Separately, I think it would be useful to update the get_status() subscription mechanism to support per-client subscriptions. I also think it would be nice if only deltas were sent after the first update. That is, if a field like gcode.gcode_position hasn't changed, then no need to send it every update. If there's interest in that, I can put together a prototype.

I think both of these are good ideas.

Arksine commented on klippy/webhooks.py:
```diff
@@ -417,31 +407,26 @@ def _process_status_request(self, requested_objects, eventtime):
+        self.add_subscripton(args)
+        cconn = web_request.get_client_connection()
+        template = web_request.get('response_template', {})
+        self.clients[cconn] = template
```
Should we only update self.clients if the `response_template` is in the web_request?  Moonraker is always going to make an initial subscription request so it can track Klippy's state, where it can set the `response_template`.  After that, Moonraker's clients will send these requests.  

I can have Moonraker add a `response_template` to requests from its clients if necessary.

KevinOConnor commented:
> When procedurally generating APIs in Moonraker 

As a separate idea - what if we allowed Moonraker clients to open a websocket directly to the Klippy webhooks server?  That is, on each creation of one of these websockets, Moonraker could open a new unix domain socket client and then just copy the contents of data between the two sockets without altering or inspecting it.

The advantage of doing that would be lower maintenance costs - we wouldn't need to document both the Klipper api and the equivalent Moonraker api.  Also, on a change, we wouldn't have to change Mainsail, Moonraker, and Klipper.  It should be a bit more efficient as well.

The disadvantage would be that clients would need to know to route some requests to Klipper (eg, motion requests, print requests, query requests) and route some requests to Moonraker (eg, file upload, software upgrade requests).

> For example, api_server/objects_subscription would resolve to a post_printer_api_server_objects_subscription websocket API call.

Renaming isn't that important.  So, if you prefer webhooks, or a different name then that's fine with me.  I think it would be useful to have a convention where we use a name prefix dependent on the module name though.  Otherwise I fear we run into naming conflicts in the future (like we're starting to hit with gcode command names now).

> We could send the function name along with the registered "method" name in list_endpoints, then moonraker could use that info to reject invalid requests. 

FWIW, I think it would be preferable to pass this type of information explicitly instead of trying to infer it from the function name.  That said, I'm not sure I understand the value in it.  There aren't that many "query" endpoints, so it would seem simple to always use a POST call.  (I understand the http convention of using GET for queries, but I don't see the value of that convention in this context.)

> I think we can remove api_server/objects_list_subscription and instead just return the current subscription data when the user adds a subscription.

Not necessarily related, but I'm a little confused on how subscriptions are intended to work.  If moonraker itself subscribes to some items, and mainsail subscribes to some items, and some other http client subscribes to some items..   How will moonraker know where to forward responses?  What if one of the client's disconnects and does not return - what happens to its subscriptions?

-Kevin

Arksine commented:
>As a separate idea - what if we allowed Moonraker clients to open a websocket directly to the Klippy webhooks server? That is, on each creation of one of these websockets, Moonraker could open a new unix domain socket client and then just copy the contents of data between the two sockets without altering or inspecting it.

The websocket implements JSON-RPC 2.0, so to directly pass requests from websocket clients to Klipper we would either have to implement the full JSON-RPC 2.0 protocol in Klipper or ask Meteyou to change protocol in Mainsail.  Since I have to inspect requests to route them to the correct application I don't think its worth changing the protocol.  That said, now that  Klippy no longer allows an endpoint to handle both GET and POST methods I could simply use the existing Klippy APIs for the websocket rather than generate new ones.  Also, I think there might be still be some value in opening a unique Unix Domain Socket for each websocket with regard to subscriptions.

>Renaming isn't that important. So, if you prefer webhooks, or a different name then that's fine with me. I think it would be useful to have a convention where we use a name prefix dependent on the module name though. Otherwise I fear we run into naming conflicts in the future (like we're starting to hit with gcode command names now).

I agree in principle.   My thought was that webhooks, or api_server, is the remote entry point into Klipper, so it made sense for its web APIs to be directly accessible via "/printer", for example:
```
GET /printer/info
POST /printer/emergency_stop
GET /printer/objects/list
GET /printer/objects/status
POST /printer/objects/subscriptions
````
vs
```
GET /printer/api_server/info
POST /printer/api_server/emergency_stop
GET /printer/api_server/objects_list
GET /printer/api_server/objects_status
POST /printer/api_server/objects_subscriptions
````
My concern with the second is that it could be confusing for web developers.  It appears that these requests operate on the  printer's "api_server".   The other requests apply the module name, however their intention is more clear:
```
GET /printer/gcode/help
POST /printer/gcode/script
GET /printer/query_endstops/status
```
I will do it however you think is best though. 

>(I understand the http convention of using GET for queries, but I don't see the value of that convention in this context.)

It isn't strictly necessary to conform.  In this case I think it is best to simply accept both GET and POST requests as I am currently doing.

>Not necessarily related, but I'm a little confused on how subscriptions are intended to work. If moonraker itself subscribes to some items, and mainsail subscribes to some items, and some other http client subscribes to some items.. How will moonraker know where to forward responses? What if one of the client's disconnects and does not return - what happens to its subscriptions?

Currently Moonraker forwards the superset of all subscriptions to all connected clients.  Clients can filter out data that they didn't request, what is most important is that clients get the data that they requested.   If a client disconnects the subscription is still active, Moonraker will receive all subscribed objects and continue forwarding them to any remaining connected clients.   However, we could use your idea of a unique Unix Domain Socket for each websocket to change this behavior.  Each client would only get the data they requested, and when that client disconnects the UDS would be removed.

Arksine commented:
Hi Kevin, I was thinking about this a bit more, what do you think about giving endpoints in api_server.py a "klipper_host" namespace?  For example:
```python
register_endpoint("klipper_host.info", self._handle_info_request)
register_endpoint("klipper_host.emergency_stop", self._handle_estop_request)	        
register_endpoint("klipper_host.objects.list", self._handle_object_request)
register_endpoint("klipper_host.objects.status",  self._handle_status_request)
register_endpoint("klipper_host.objects.subscription",  self._handle_subscription_request)
```
The above would yield the following HTTP APIs:
```
/printer/klipper_host/info
/printer/klipper_host/emergency_stop
/printer/klipper_host/objects/list
/printer/klipper_host/objects/status
/printer/klipper_host/objects/subscription
```
and the following websocket APIs:
```
printer.klipper_host.info
printer.klipper_host.emergency_stop
printer.klipper_host.objects.list
printer.klipper_host..objects.status
printer.klipper_host.objects.subscription
```

All other modules would use the module name as namespace (ie: `register_endpoint("gcode.script")`).  This approach has the advantage of providing namespaces to prevent naming conflicts and provides Web APIs that developers can infer the meaning.  

If you prefer to keep it the way it is I'm ok with it as well.  In general I think the proposal is a good move forward for the API.

KevinOConnor commented:
Hi @Arksine .  I was giving it some further thought as well and I concluded the webhooks.py code is special anyway, so it's not a big deal if it does not follow a standard prefix.  I was just planning to drop the final patch in this series.  (That is, not bother with a rename, and not bother with a consistent prefix for webhooks.py.)

> One other thing I thought of, and completely up to you, is that perhaps our naming convention for Klippy endpoints should be something like: api_server.info rather than api_server/info. I can replace the "." when I generate the HTTP APIs, and use printer.api_server.info for the websocket API.

I have no preference wrt "." or "/" - so whatever you prefer is fine with me.

Just to summarize - it sounds like we're in sync with: adding the whconsole.py test tool, only creating the unix domain socket with a new klippy command-line argument, removing register_static_path() (in favor of extending the "info" endpoint), removing call_remote_method() (in favor of making "objects/subscribe" and "gcode output" subscription based), and converting the protocol to a more json-rpc format?  Also, I'll drop the api rename for now, and I'll take a look at a prototype for per-client subscriptions with delta transmission?  I'm okay with adding a `webhooks.register_endpoint(..., is_query_only=True)` if you think that's useful.

If I got any of that wrong, just let me know.

Cheers,
-Kevin

Arksine commented:
>Just to summarize - it sounds like we're in sync with: adding the whconsole.py test tool, only creating the unix domain socket with a new klippy command-line argument, removing register_static_path() (in favor of extending the "info" endpoint), removing call_remote_method() (in favor of making "objects/subscribe" and "gcode output" subscription based), and converting the protocol to a more json-rpc format? Also, I'll drop the api rename for now, and I'll take a look at a prototype for per-client subscriptions with delta transmission? 

All of that sounds good to me.

> I have no preference wrt "." or "/" - so whatever you prefer is fine with me.

I'm okay either way, so we can leave it as "/".  

> I'm okay with adding a webhooks.register_endpoint(..., is_query_only=True) if you think that's useful.

I think you are correct that in this context it isn't important to filter request methods, so accepting both GET and POST for requests to Klippy is the best solution.

KevinOConnor commented:
FYI, I've updated this branch.  Notable changes:
1. I've dropped the api_server rename.
2. I've moved the gcode webhooks code from gcode.py to webhooks.py and the gcode.py code is now initialized before webhooks.py.  (The endpoint names were kept the same though.)
3. I've put up some prototype code for a new "query get_status()" system.

The new get_status() code only has two endpoints: `objects/list` and `objects/subscribe`.  An example subscribe might look like:
`{"method": "objects/subscribe", "params": {"objects": {"toolhead": ["status"], "webhooks": null}}}`
and an initial response might look like:
`{"params": {"status": {"webhooks": {"state": "startup", "state_message": "\nPrinter is not ready\nThe klippy host software is attempting to connect.  Please\nretry in a few moments.\n"}, "toolhead": {"status": "Printing"}}, "eventtime": 5605708.720735481}}`
Later on, one might get an update like:
`{"params": {"status": {"webhooks": {"state": "ready", "state_message": "Printer is ready"}}, "eventtime": 5605747.564345255}}`
After the initial subscription, only changes are sent.  If there are no changes at all, then no message is sent.

Thoughts?
-Kevin

Arksine commented:
I think it looks good.  Notable API changes are that subscriptions are sent within an `objects` parameter, and when a client wishes to subscribe to all items of an object the value should be `null` value rather than an empty list.  Both make sense to me.

Would it be possible to add the ability to poll status via an endpoint and send the data back in the response?  This is useful for clients that wish to use the HTTP API exclusively.  Specifically, I expect to see "clients" running on ESP32 devices that only use http requests.  Since these clients have no websocket they would have no way of retrieving state.

KevinOConnor commented:
> Would it be possible to add the ability to poll status via an endpoint and send the data back in the response? This is useful for clients that wish to use the HTTP API exclusively.

Okay, that makes sense.  I reworked the code so one can issue an "objects/query" command that will work like "objects/subscribe", but wont provide any further updates after the initial query.  I also changed the subscribe command so that the initial snapshot comes as part of the response, and the asynchronous messages only ever contain incremental updates.

Not sure if the new code is as clear as it could be however.  Does the API look okay though?

-Kevin

Arksine commented:
The API is good, an improvement on the previous IMO. 

I am currently making the changes to Moonraker to test against this branch, once that is done I will get with meteyou and he can prepare the next release for Mainsail that will be compatible with these changes.  I really like the diffs, I think it will be a significant improvement considering the amount of status that likely doesn't change often.

Arksine commented:
Hi Kevin,
I have this branch working with Moonraker, all is good.  I think the only thing we need to do is return the "shutdown" state in klippy.py:
```diff
--- a/klippy/klippy.py
+++ b/klippy/klippy.py
@@ -66,7 +66,9 @@ class Printer:
     def get_reactor(self):
         return self.reactor
     def get_state_message(self):
-        if self.state_message == message_ready:
+        if self.in_shutdown_state:
+            category = "shutdown"
+        elif self.state_message == message_ready:
             category = "ready"
         elif self.state_message == message_startup:
             category = "startup"
```

This makes it a little bit easier for clients to detect a shutdown via subscription to "webhooks".

KevinOConnor commented:
Great!  I went ahead and merged this branch.

-Kevin
